### PR TITLE
Web3Signer: CLI implementation

### PR DESCRIPTION
**Pull request by james-prysm**
**What type of PR is this?**

Feature

Part of #9994

**What does this PR do? Why is it needed?**
This PR introduces new CLI commands for the web3 signer keymanager type and integrates the keymanager into our runtime. The Web3Signer should be functional after this PR.

### CODE REVIEW

File: cmd/validator/flags/flags.go [hunk 1]
```diff
@@ -267,6 +267,25 @@ var (
 		Usage: "/path/to/ca.crt for establishing a secure, TLS gRPC connection to a remote signer server",
 		Value: "",
 	}
+	// Web3SignerURLFlag defines the URL for a web3signer to connect to.
+	// example:--validators-external-signer-url=http://localhost:9000
+	// web3signer documentation: https://docs.web3signer.consensys.net/en/latest/Tutorials/Launchpad-Keystores/
```
[the remark this answers is not in the document]
Comment (james-prysm) [reply cmd/validator/flags/flags.go:272]: will change

File: cmd/validator/flags/flags.go [hunk 2]
```diff
@@ -267,6 +267,25 @@ var (
 		Usage: "/path/to/ca.crt for establishing a secure, TLS gRPC connection to a remote signer server",
 		Value: "",
 	}
+	// Web3SignerURLFlag defines the URL for a web3signer to connect to.
+	// example:--validators-external-signer-url=http://localhost:9000
+	// web3signer documentation: google Consensys' web3signer docs
+	Web3SignerURLFlag = &cli.StringFlag{
+		Name:  "validators-external-signer-url",
+		Usage: "webhook for web3signer to use for validator signing, please enter the base url of the web3signer",
```
Comment (rauljordan) [inline cmd/validator/flags/flags.go:275]:
```suggestion
		Usage: "URL for consensys' web3signer software to use with the Prysm validator client",
```

File: cmd/validator/flags/flags.go [hunk 3]
```diff
@@ -267,6 +267,25 @@ var (
...
+		Value: "",
+	}
+
+	// Web3SignerPublicValidatorKeysFlag defines a comma-separated list of hex string public keys or external url for web3signer to use for validator signing.
+	// example with external url: --validators-external-signer-public-keys= https://web3signer.com/api/v1/eth2/publicKeys
+	// example with public key: --validators-external-signer-public-keys=0xa99a...e44c,0xb89b...4a0b
+	// web3signer documentation: google Consensys' web3signer docs
+	Web3SignerPublicValidatorKeysFlag = &cli.StringFlag{
+		Name:  "validators-external-signer-public-keys",
+		Usage: "comma separated list of public keys or eternal url endpoint for validators to use for signing, for web3signer",
```
Comment (rauljordan) [inline cmd/validator/flags/flags.go:285]:
```suggestion
		Usage: "comma separated list of public keys OR an external url endpoint for the validator to retrieve public keys from for usage with web3signer",
```

File: cmd/validator/main.go [hunk 4]
```diff
@@ -64,16 +64,20 @@ var appFlags = []cli.Flag{
...
 	flags.MonitoringPortFlag,
-	cmd.DisableMonitoringFlag,
 	flags.SlasherRPCProviderFlag,
 	flags.SlasherCertFlag,
 	flags.WalletPasswordFileFlag,
 	flags.WalletDirFlag,
 	flags.EnableWebFlag,
 	flags.GraffitiFileFlag,
 	flags.EnableDutyCountDown,
+	// web3signer flags
```
Comment (rauljordan) [inline cmd/validator/main.go:75]:
```suggestion
	// Consensys' Web3Signer flags
```

File: cmd/validator/wallet/wallet.go [hunk 5]
```diff
@@ -30,6 +30,10 @@ var Commands = &cli.Command{
 				flags.RemoteSignerCertPathFlag,
 				flags.RemoteSignerKeyPathFlag,
 				flags.RemoteSignerCACertPathFlag,
+
```
Comment (rauljordan) [inline cmd/validator/wallet/wallet.go:33]:
```suggestion
```

File: cmd/validator/wallet/wallet.go [hunk 6]
```diff
@@ -30,6 +30,10 @@ var Commands = &cli.Command{
...
+				flags.Web3SignerURLFlag,
+				flags.Web3SignerPublicValidatorKeysFlag,
+
```
Comment (rauljordan) [inline cmd/validator/wallet/wallet.go:36]:
```suggestion
```

File: cmd/validator/wallet/wallet.go [hunk 7]
```diff
@@ -63,6 +67,10 @@ var Commands = &cli.Command{
 				flags.RemoteSignerCertPathFlag,
 				flags.RemoteSignerKeyPathFlag,
 				flags.RemoteSignerCACertPathFlag,
+
```
Comment (rauljordan) [inline cmd/validator/wallet/wallet.go:70]:
```suggestion
```

File: cmd/validator/wallet/wallet.go [hunk 8]
```diff
@@ -63,6 +67,10 @@ var Commands = &cli.Command{
...
+				flags.Web3SignerURLFlag,
+				flags.Web3SignerPublicValidatorKeysFlag,
+
```
Comment (rauljordan) [inline cmd/validator/wallet/wallet.go:73]:
```suggestion
```

File: validator/accounts/accounts_backup.go [hunk 9]
```diff
@@ -48,9 +48,9 @@ func BackupAccountsCli(cliCtx *cli.Context) error {
 	if err != nil {
 		return errors.Wrap(err, "could not initialize wallet")
 	}
-	if w.KeymanagerKind() == keymanager.Remote {
+	if w.KeymanagerKind() == keymanager.Remote || w.KeymanagerKind() == keymanager.Web3Signer {
```
Comment (rauljordan) [inline validator/accounts/accounts_backup.go:51]: the long-term solution is to have an interface called `KeyBackuper` and check if the keymanager implements that interface, instead of all these checks for what it is not. No need to do so in this PR, as we can do that cleanup later with an external contributor that is interested in it. @michaelneuder FYI

File: validator/accounts/wallet/wallet.go [hunk 10]
```diff
@@ -290,6 +292,25 @@ func (w *Wallet) InitializeKeymanager(ctx context.Context, cfg iface.InitKeymana
...
+	case keymanager.Web3Signer: //option should not work for --web
+		configFile, err := w.ReadKeymanagerConfigFromDisk(ctx)
+		if err != nil {
+			return nil, errors.Wrap(err, "could not read web3signer keymanager config")
+		}
+		config, err := remote_web3signer.UnmarshalConfigFile(configFile)
+		if err != nil {
+			return nil, errors.Wrap(err, "could not unmarshal web3signer keymanager config file")
+		}
+		// injected in the validator runner
```
Comment (rauljordan) [inline validator/accounts/wallet/wallet.go:304]:
```suggestion
```

File: validator/accounts/wallet/wallet.go [hunk 11]
```diff
@@ -290,6 +292,25 @@ func (w *Wallet) InitializeKeymanager(ctx context.Context, cfg iface.InitKeymana
...
+		if len(cfg.GenesisValidatorsRoot) != 0 {
```
Comment (rauljordan) [inline validator/accounts/wallet/wallet.go:305]: Simplify by checking if len(cfg.GenesisValidatorsRoot) != 32 { return nil, errors.New("genesis validators root value is required for web3signer keymanager but was not provided" }

File: validator/client/aggregate.go [hunk 12]
```diff
@@ -5,6 +5,8 @@ import (
 	"fmt"
 	"time"
 
+	"github.com/prysmaticlabs/prysm/network/forks"
+
```
Comment (rauljordan) [inline validator/client/aggregate.go:9]:
```suggestion
```

File: validator/client/aggregate.go [hunk 13]
```diff
@@ -129,6 +131,10 @@ func (v *validator) signSlotWithSelectionProof(ctx context.Context, pubKey [fiel
 	if err != nil {
 		return nil, err
 	}
+	fork, err := forks.Fork(slots.ToEpoch(slot))
```
Comment (rauljordan) [inline validator/client/aggregate.go:134]: Do you still need this?

File: validator/client/sync_committee.go [hunk 14]
```diff
@@ -4,6 +4,8 @@ import (
 	"context"
 	"fmt"
 
+	"github.com/prysmaticlabs/prysm/network/forks"
+
```
Comment (rauljordan) [inline validator/client/sync_committee.go:8]:
```suggestion
```

File: validator/client/aggregate.go (hunk 13 above)
Comment (james-prysm) [reply validator/client/aggregate.go:133]: nope must have missed in merge conflicts still in the process of cleaning that up after breaking up the prs.

File: validator/accounts/wallet/wallet.go [hunk 15]
```diff
@@ -290,6 +293,26 @@ func (w *Wallet) InitializeKeymanager(ctx context.Context, cfg iface.InitKeymana
...
+		if err != nil {
+			return nil, errors.Wrap(err, "could not unmarshal web3signer keymanager config file")
+		}
+		// TODO(9883): future work needs to address how initialize keymanager is called for web3signer.
+		// an error may be thrown for genesis validators root for some InitializeKeymanager calls.
+		if len(cfg.GenesisValidatorsRoot) != fieldparams.RootLength {
+			config.GenesisValidatorsRoot = cfg.GenesisValidatorsRoot
+		} else {
+			return nil, errors.New("could not set genesis validators root for web3signer keymanager")
+		}
```
Comment (rauljordan) [inline validator/accounts/wallet/wallet.go:311]: This should just throw an error like this:
```go
if len(cfg.GenesisValidatorsRoot) != 32 {
  return nil, errors.New("web3signer requires a genesis validators root value")
}
```

File: validator/node/node.go [hunk 16]
```diff
@@ -282,17 +268,23 @@ func (c *ValidatorClient) initializeForWeb(cliCtx *cli.Context) error {
...
+	//	}
+	//	log.WithFields(logrus.Fields{
+	//		"wallet":          w.AccountsDir(),
+	//		"keymanager-kind": w.KeymanagerKind().String(),
+	//	}).Info("Opened validator wallet")
+	//	keyManager, err = w.InitializeKeymanager(cliCtx.Context, accountsiface.InitKeymanagerConfig{ListenForChanges: true})
+	//	if err != nil {
+	//		return errors.Wrap(err, "could not read keymanager for wallet")
+	//	}
+	//}
```
Comment (rauljordan) [inline validator/node/node.go:287]: Remove commented code

File: validator/rpc/standard_api_test.go [hunk 17]
```diff
@@ -102,11 +101,10 @@ func TestServer_ImportKeystores(t *testing.T) {
 		SkipMnemonicConfirm: true,
 	})
 	require.NoError(t, err)
-	km, err := w.InitializeKeymanager(ctx, iface.InitKeymanagerConfig{ListenForChanges: false})
-	require.NoError(t, err)
+	//km, err := w.InitializeKeymanager(ctx, iface.InitKeymanagerConfig{ListenForChanges: false})
+	//require.NoError(t, err)
```
Comment (rauljordan) [inline validator/rpc/standard_api_test.go:105]: Remove commented code

File: validator/rpc/standard_api_test.go [hunk 18]
```diff
@@ -357,11 +357,10 @@ func setupServerWithWallet(t testing.TB) *Server {
 		SkipMnemonicConfirm: true,
 	})
 	require.NoError(t, err)
-	km, err := w.InitializeKeymanager(ctx, iface.InitKeymanagerConfig{ListenForChanges: false})
-	require.NoError(t, err)
+	//km, err := w.InitializeKeymanager(ctx, iface.InitKeymanagerConfig{ListenForChanges: false})
```
Comment (rauljordan) [inline validator/rpc/standard_api_test.go:360]: +1

File: cmd/validator/flags/flags.go [hunk 19]
```diff
@@ -267,6 +267,25 @@ var (
 		Usage: "/path/to/ca.crt for establishing a secure, TLS gRPC connection to a remote signer server",
 		Value: "",
 	}
+	// Web3SignerURLFlag defines the URL for a web3signer to connect to.
+	// example:--validators-external-signer-url=http://localhost:9000
+	// web3signer documentation: google Consensys' web3signer docs
```
Comment (rauljordan) [inline cmd/validator/flags/flags.go:272]:
```suggestion
	// web3signer documentation can be found in Consensys' web3signer project docs
```

File: cmd/validator/flags/flags.go [hunk 20]
```diff
@@ -267,6 +267,25 @@ var (
...
+	Web3SignerURLFlag = &cli.StringFlag{
+		Name:  "validators-external-signer-url",
+		Usage: "URL for consensys' web3signer software to use with the Prysm validator client",
+		Value: "",
+	}
+
+	// Web3SignerPublicValidatorKeysFlag defines a comma-separated list of hex string public keys or external url for web3signer to use for validator signing.
+	// example with external url: --validators-external-signer-public-keys= https://web3signer.com/api/v1/eth2/publicKeys
+	// example with public key: --validators-external-signer-public-keys=0xa99a...e44c,0xb89b...4a0b
+	// web3signer documentation: google Consensys' web3signer docs
```
Comment (rauljordan) [inline cmd/validator/flags/flags.go:282]:
```suggestion
	// web3signer documentation can be found in Consensys' web3signer project docs```

File: validator/client/service.go [hunk 21]
```diff
@@ -228,38 +228,23 @@ func (v *ValidatorService) Status() error {
...
-	go recheckValidatingKeysBucket(ctx, v.db, v.keyManager)
-	for _, key := range validatingKeys {
-		log.WithField(
-			"publicKey", fmt.Sprintf("%#x", bytesutil.Trunc(key[:])),
-		).Info("Validating for public key")
-	}
+///////////////////////////////////////////////////////////////////////////////
+// VlidatorService Getters and Setters ////////////////////////////////////////
+///////////////////////////////////////////////////////////////////////////////
+
```
Comment (rauljordan) [inline validator/client/service.go:234]:
```suggestion

```

File: validator/client/service.go [hunk 22]
```diff
@@ -228,38 +228,23 @@ func (v *ValidatorService) Status() error {
...
+	return v.interopKeysConfig
+}
+
+func (v *ValidatorService) Keymanager() (keymanager.IKeymanager, error) {
+	return v.validator.Keymanager()
 }
 
+///////////////////////////////////////////////////////////////////////////////
+///////////////////////////////////////////////////////////////////////////////
+///////////////////////////////////////////////////////////////////////////////
```
Comment (rauljordan) [inline validator/client/service.go:246]: No need for these separators in general

Comment (rauljordan) [inline validator/client/service.go:246]:
```suggestion
```

**Review by rauljordan — DISMISSED** [review]

**Comment by prestonvanloon** [comment]
Please enable e2e tests for web3signer before merging this

File: testing/endtoend/components/validator.go [hunk 23]
```diff
@@ -161,8 +160,7 @@ func (v *ValidatorNode) Start(ctx context.Context) error {
 		for _, pub := range pubs {
 			hexPubs = append(hexPubs, "0x"+hex.EncodeToString(pub.Marshal()))
 		}
-		// TODO(9994): Replace "validators-external-signer-public-keys" with flags.Web3RemoteSignerPubkeysFlag.Name
-		args = append(args, fmt.Sprintf("--validators-external-signer-public-keys=%s", strings.Join(hexPubs, ",")))
+		args = append(args, fmt.Sprintf("--%s=%s", flags.Web3SignerPublicValidatorKeysFlag, strings.Join(hexPubs, ",")))
```
Comment (prestonvanloon) [inline testing/endtoend/components/validator.go:163]:
```suggestion
		args = append(args, fmt.Sprintf("--%s=%s", flags.Web3SignerPublicValidatorKeysFlag.Name, strings.Join(hexPubs, ",")))
```

File: testing/endtoend/components/validator.go [hunk 24]
```diff
@@ -149,8 +149,7 @@ func (v *ValidatorNode) Start(ctx context.Context) error {
 		args = append(args, features.E2EValidatorFlags...)
 	}
 	if v.config.UseWeb3RemoteSigner {
-		// TODO(9994): Replace "validators-external-signer-url" with flags.Web3RemoteSignerURLFlag.Name
-		args = append(args, fmt.Sprintf("--%s=localhost:%d", "validators-external-signer-url", Web3RemoteSignerPort))
+		args = append(args, fmt.Sprintf("--%s=localhost:%d", flags.Web3SignerURLFlag, Web3RemoteSignerPort))
```
Comment (prestonvanloon) [inline testing/endtoend/components/validator.go:152]:
```suggestion
		args = append(args, fmt.Sprintf("--%s=localhost:%d", flags.Web3SignerURLFlag.Name, Web3RemoteSignerPort))
```

**Comment by prestonvanloon** [comment]
OK to skip e2e tests pending https://github.com/ConsenSys/teku/issues/4905 and https://github.com/prysmaticlabs/prysm/pull/10123

File: validator/keymanager/remote-web3signer/internal/client.go [hunk 25]
```diff
@@ -0,0 +1,179 @@
...
+
+type SignRequestJson []byte
+
+// HttpSignerClient defines the interface for interacting with a remote web3signer.
+type HttpSignerClient interface {
+	Sign(ctx context.Context, pubKey string, request SignRequestJson) (bls.Signature, error)
+	GetPublicKeys(ctx context.Context, url string) ([][48]byte, error)
+}
+
+// ApiClient a wrapper object around web3signer APIs. API docs found here https://consensys.github.io/web3signer/web3signer-eth2.html.
```
Comment (rauljordan) [inline validator/keymanager/remote-web3signer/internal/client.go:34]: Concerned about link becoming broken in the future. Maybe we can just say refer to the docs from Consensys' web3signer project

**Review by rauljordan — APPROVED** [review]

File: validator/keymanager/remote-web3signer/internal/client.go [hunk 26]
```diff
@@ -0,0 +1,179 @@
...
+// NewApiClient method instantiates a new ApiClient object.
+func NewApiClient(baseEndpoint string) (*ApiClient, error) {
+	u, err := url.Parse(baseEndpoint)
+	if err != nil {
+		return nil, errors.Wrap(err, "invalid format, unable to parse url")
+	}
+	return &ApiClient{
+		BaseURL: u,
+		RestClient: &http.Client{
+			Timeout: maxTimeout,
```
Comment (prestonvanloon) [inline validator/keymanager/remote-web3signer/internal/client.go:49]: Let's remove this timeout. All requests require a context.Context, which the caller sets the deadline/timeout on the request. There is no need to set a very short 3s timeout when the caller has the ability to do this. Most of these contexts have a deadline set to the end of the slot anyway.

File: validator/rpc/accounts.go [hunk 27]
```diff
@@ -179,16 +189,27 @@ func (s *Server) DeleteAccounts(
...
+	if err != nil {
+		return nil, err
+	}
+	switch ikm := ikm.(type) {
+	case *imported.Keymanager:
+		km = ikm
+	case *derived.Keymanager:
+		km = ikm
+	default:
+		return nil, status.Error(codes.FailedPrecondition, "Only Imported or Derived wallets can delete accounts")
```
Comment (prestonvanloon) [inline validator/rpc/accounts.go:212]: I dont understand this. Why can't a remote signer perform a voluntary exit?

File: validator/keymanager/remote-web3signer/keymanager.go [hunk 28]
```diff
@@ -106,32 +108,33 @@ func getSignRequestJson(request *validatorpb.SignRequest, genesisValidatorsRoot
 		if err != nil {
 			return nil, err
 		}
-		return json.Marshal(bockSignRequest)
+		return json.MarshalIndent(bockSignRequest, "", "\t")
```
Comment (prestonvanloon) [inline validator/keymanager/remote-web3signer/keymanager.go:111]: Why do we need to pretty print this?

File: validator/accounts/wallet/wallet.go [hunk 29]
```diff
@@ -274,7 +290,7 @@ func (w *Wallet) InitializeKeymanager(ctx context.Context, cfg iface.InitKeymana
 		if err != nil {
 			return nil, errors.Wrap(err, "could not initialize derived keymanager")
 		}
-	case keymanager.Remote:
+	case keymanager.Remote: //option should not work for --web
```
Comment (prestonvanloon) [inline validator/accounts/wallet/wallet.go:293]: What's up with this comment?

File: validator/accounts/wallet/wallet.go [hunk 30]
```diff
@@ -290,6 +306,20 @@ func (w *Wallet) InitializeKeymanager(ctx context.Context, cfg iface.InitKeymana
 		if err != nil {
 			return nil, errors.Wrap(err, "could not initialize remote keymanager")
 		}
+	case keymanager.Web3Signer: //option should not work for --web
```
Comment (prestonvanloon) [inline validator/accounts/wallet/wallet.go:309]: Same here, what's up with this comment?

File: validator/accounts/wallet_edit.go [hunk 31]
```diff
@@ -49,6 +49,8 @@ func EditWalletConfigurationCli(cliCtx *cli.Context) error {
 		if err := w.WriteKeymanagerConfigToDisk(cliCtx.Context, encodedCfg); err != nil {
 			return errors.Wrap(err, "could not write config to disk")
 		}
+	case keymanager.Web3Signer:
+		return errors.New("web3signer keymanager does not support editing configuration or persistent prysm wallet")
```
Comment (prestonvanloon) [inline validator/accounts/wallet_edit.go:53]: I think you can remove this. It will be captured in the default with a defined error for not supported key manager. 

```suggestion
```

File: validator/client/runner.go [hunk 32]
```diff
@@ -35,11 +35,7 @@ var backOffPeriod = 10 * time.Second
 func run(ctx context.Context, v iface.Validator) {
 	cleanup := v.Done
 	defer cleanup()
-	if err := v.WaitForWalletInitialization(ctx); err != nil {
-		// log.Fatalf will prevent defer from being called
-		cleanup()
-		log.Fatalf("Wallet is not ready: %v", err)
-	}
+
```
Comment (prestonvanloon) [inline validator/client/runner.go:38]: Why is this moved?

File: validator/keymanager/remote-web3signer/keymanager_test.go [hunk 33]
```diff
@@ -269,23 +267,28 @@ func TestKeymanager_FetchValidatingPublicKeys_WithExternalURL_ThrowsError(t *tes
...
-	_, err = buffer.Write(b)
-	require.NoError(t, err)
-	r := ioutil.NopCloser(&buffer)
-
-	config, err := UnmarshalConfigFile(r)
+func TestPrettyPrintSignRequest_HappyPath(t *testing.T) {
+	root, err := hexutil.Decode("0x270d43e74ce340de4bca2b1936beca0f4f5408d9e78aec4850920baf659d5b69")
 	assert.NoError(t, err)
-	assert.Equal(t, fakeConfig.BaseEndpoint, config.BaseEndpoint)
+	fmt.Printf("%v", root)
```
Comment (prestonvanloon) [inline validator/keymanager/remote-web3signer/keymanager_test.go:273]:
```suggestion
```

**Review by prestonvanloon — CHANGES_REQUESTED** [review]

File: validator/accounts/wallet/wallet.go (hunk 29 above)
Comment (james-prysm) [reply validator/accounts/wallet/wallet.go:293]: the only types that work with the UI are keymanager types that are native to prysm. the UI doesn't support remote signer types as it defeats the purpose of onboarding ( importing your keys to prysm) and creating the prysm wallet.

File: validator/accounts/wallet/wallet.go (hunk 30 above)
Comment (james-prysm) [reply validator/accounts/wallet/wallet.go:309]: same as the above response

File: validator/client/runner.go (hunk 32 above)
Comment (james-prysm) [reply validator/client/runner.go:38]: the web3signer requires the genesis validators root that only happens after chain start

File: validator/rpc/accounts.go (hunk 27 above)
Comment (james-prysm) [reply validator/rpc/accounts.go]: some old code threw me off I don't understand why the current remote signer can't voluntarily exit. i made change to remove this restriction

File: validator/accounts/wallet_edit.go (hunk 31 above)
Comment (prestonvanloon) [reply validator/accounts/wallet_edit.go]: Why was this resolved?

File: validator/keymanager/remote-web3signer/keymanager.go (hunk 28 above)
Comment (james-prysm) [reply validator/keymanager/remote-web3signer/keymanager.go]: reverted this

File: WORKSPACE [hunk 34]
```diff
@@ -397,7 +397,49 @@ load("@com_github_atlassian_bazel_tools//gometalinter:deps.bzl", "gometalinter_d
...
+    sum = "h1:BqpAaACuzVSgi/VLzGZIobT2z4v53pjosyNd9Yv6n/w=",
+    version = "v1.2.1",
+)
+
+go_repository(
+    name = "com_github_pkg_diff",
+    importpath = "github.com/pkg/diff",
+    sum = "h1:aoZm08cpOy4WuID//EZDgcC4zIxODThtZNPirFr42+A=",
+    version = "v0.0.0-20210226163009-20ebb0f2a09e",
+)
```
Comment (prestonvanloon) [inline WORKSPACE:442]: Do not add go dependencies to WORKSPACE. See deps.bzl

File: validator/client/aggregate.go [hunk 35]
```diff
@@ -183,6 +183,7 @@ func (v *validator) aggregateAndProofSig(ctx context.Context, pubKey [fieldparam
 	if err != nil {
 		return nil, err
 	}
+	log.Warnf("ATTEMPTING TO SIGN AGGREGATE AND PROOF")
```
Comment (rauljordan) [inline validator/client/aggregate.go:186]:
```suggestion
```

File: validator/client/aggregate.go [hunk 36]
```diff
@@ -191,6 +192,7 @@ func (v *validator) aggregateAndProofSig(ctx context.Context, pubKey [fieldparam
 		SigningSlot:     slot,
 	})
 	if err != nil {
+		log.Warnf("GOT ERROR FROM SIGN FUNCTION FOR AGGREGATE AND PROOF")
```
Comment (rauljordan) [inline validator/client/aggregate.go:195]:
```suggestion
```

File: validator/keymanager/remote-web3signer/keymanager.go [hunk 37]
```diff
@@ -36,34 +38,36 @@ type SetupConfig struct {
...
 	providedPublicKeys    [][48]byte
 	accountsChangedFeed   *event.Feed
+	validator             *validator.Validate
 }
 
 // NewKeymanager instantiates a new web3signer key manager.
 func NewKeymanager(_ context.Context, cfg *SetupConfig) (*Keymanager, error) {
-	if cfg.BaseEndpoint == "" || len(cfg.GenesisValidatorsRoot) == 0 {
-		return nil, fmt.Errorf("invalid setup config, one or more configs are empty: BaseEndpoint: %v, GenesisValidatorsRoot: %v", cfg.BaseEndpoint, cfg.GenesisValidatorsRoot)
+	if cfg.BaseEndpoint == "" || len(cfg.GenesisValidatorsRoot) != fieldparams.RootLength || bytes.Equal(params.BeaconConfig().ZeroHash[:], cfg.GenesisValidatorsRoot) {
```
Comment (rauljordan) [inline validator/keymanager/remote-web3signer/keymanager.go:51]: We perform this validation in a few places. If we only do this in this package, maybe we can turn this validation into a helper function that is unexported. If we use it in several packages, we can find a better place to put that helper function

File: encoding/bytesutil/bytes_test.go [hunk 38]
```diff
@@ -516,3 +517,9 @@ func TestSafeCopy2d32Bytes(t *testing.T) {
 	assert.Equal(t, false, &input == &output, "No copy was made")
 	assert.DeepEqual(t, input, output)
 }
+
+func TestNonZeroRoot(t *testing.T) {
+	input := make([]byte, fieldparams.RootLength)
+	output := bytesutil.NonZeroRoot(input)
+	assert.Equal(t, false, output)
+}
```
Comment (prestonvanloon) [inline encoding/bytesutil/bytes_test.go:525]: Test the true case? This method could be implemented like this and still pass your test
```go
func NonZeroRoot(root []byte) bool {
  return false
}
```

File: validator/keymanager/remote-web3signer/keymanager.go (hunk 37 above)
Comment (james-prysm) [reply validator/keymanager/remote-web3signer/keymanager.go]: create new function in bytesutil

File: encoding/bytesutil/bytes_test.go (hunk 38 above)
Comment (james-prysm) [reply encoding/bytesutil/bytes_test.go]: added cases for both false and true

File: validator/keymanager/remote-web3signer/v1/custom_mappers.go [hunk 39]
```diff
@@ -57,8 +57,8 @@ func MapAttestation(attestation *ethpb.Attestation) (*Attestation, error) {
 		return nil, err
 	}
 	return &Attestation{
```
Comment (prestonvanloon) [inline validator/keymanager/remote-web3signer/v1/custom_mappers.go:59]: Please add regression tests for removal of `.Bytes()` as this is fundamentally very different data.

**Review by prestonvanloon — APPROVED** [review]
